### ISSUE TITLE
[BUG] RSOD/Crash when kernel writes to userspace memory marked CoW

### ISSUE THREAD

**Issue by pawlos** [byteduck/duckOS#61]
## Information

Crash when closing the Viewer for the 2nd time (that's right - the first time nothing unexpected happens) non-image file in Viewer.

## Logs

```
Whoops! Something terrible happened.
If you weren't expecting this, please open an issue on GitHub to report it.
Here are the details:
SYSCALL_PAGEFAULT
A page fault occurred in the kernel (pid: 7, tid: 7, ptr: 0xc1737ffc, ip: 0xc0131c35).
```

## Screenshot

![image](https://user-images.githubusercontent.com/1296768/232554279-8dd43097-305b-4582-8bc2-001fd1050446.png)

## Environment

Describe what you were doing when the crash or bug occurred. Additionally, provide information about the environment (if you aren't sure, don't worry if you need to leave something blank):
- Win10
- Qemu 6.2.0

## To Reproduce

Steps to reproduce:
- open Viewer and select a binary file (non-image I guess)
- Viewer will report correctly that this image is not an image
- Close the Viewer
- Open the Viewer for the second time
- select again the binary file
- Viewer will report correctly that this image is not an image
- Close the Viewer -> RSOD

## Additional context
N/A

**Comment by byteduck** [comment]
Interesting. I am able to reproduce this - I had noticed similar panics from time to time, but was never able to reliably reproduce them - this seems to be a good way of doing so, and maybe I can finally get to the bottom of this.

PS: You can get full stack traces for kernel panics by passing `-DADD_KERNEL_DEBUG_SYMBOLS:BOOL=TRUE` to cmake :)

**Comment by pawlos** [comment]
@byteduck thanks for the tip with KERNEL_DEBUG_SYMBOLS. I'll use that flag next time I encounter something unexpected

**Comment by byteduck** [comment]
After a bit of a hiatus, I had some time to come back and try to debug this a bit. The issue boils down to this (and can occur in a lot of other situations):

1. When an application is closed, `SIGCHLD` is sent to Sandbar. Since Sandbar has a handler for this signal, a stack for signal handling is allocated the first time this happens.
2. Another application is opened, which involves forking Sandbar and `exec`ing. *This marks all pages mapped in Sandbar, including that signal handling stack, as CoW.*
3. The application is closed, sending `SIGCHLD` to Sandbar again. A new signal handling stack isn't allocated, since it was allocated the last time. However, all of the pages in it are still marked CoW since they haven't been touched since forking off that new application.
4. The kernel attempts to setup the signal handling stack by mapping it into kernel space. However, since the pages are still marked CoW, they are mapped read-only. This is normal behavior, as we'd want to copy the pages upon writing. However, we always panic when a pagefault occurs from code in kernel space, so we're greeted with a nice red screen and a wall of debug information instead of the CoW being handled gracefully.

I at least know why this happens now, and I'm sure there are other situations in which something like this may happen that I didn't think of (for example, some syscall that writes to memory in userspace that may be marked CoW). I'll have to update the kernel to handle this case. Additionally, it may be a good idea to just unmark pages as CoW if only one process is using them anyway, so we avoid unnecessary copies.

**Comment by byteduck** [comment]
Fixed in d62b7c8 by simply not including the signal handler stack region when forking.
